File: apps/backend/agent-api/api/schemas/generate.py
from pydantic import BaseModel
from typing import Optional, Literal


class Message(BaseModel):
    role: str
    content: str


class ContentOverview(BaseModel):
    learningObjectives: list[str]
    keyConcepts: list[str]


ContentType = Literal["article", "slide", "quiz", "flashcard"]


class TopicInput(BaseModel):
    title: str
    contentOverview: Optional[ContentOverview] = None
    contentType: Optional[ContentType] = None


class GenerateTopicsRequest(BaseModel):
    topic: str
    goal: str
    messages: Optional[list[Message]] = None


class GenerateTopicsResponse(BaseModel):
    title: str
    topics: list[TopicInput]


class ChatMessageRequest(BaseModel):
    message: str
    phase: str
    topic: Optional[str] = None
    goal: Optional[str] = None
    topics: Optional[list[TopicInput]] = None
    messages: Optional[list[Message]] = None


class ChatMessageResponse(BaseModel):
    intent: str
    response: str
    data: Optional[dict] = None


class GenerateTopicsFromChatRequest(BaseModel):
    messages: list[Message]
    content_type: ContentType

File: apps/frontend/web/src/api/generate.ts
export interface Message {
  role: 'user' | 'assistant'
  content: string
}

export interface GenerateTopicsRequest {
  topic: string
  goal: string
  messages?: Message[]
}

export interface ContentOverview {
  learningObjectives: string[]
  keyConcepts: string[]
}

export type ContentType = 'article' | 'slide' | 'quiz' | 'flashcard'

export interface TopicInput {
  title: string
  contentOverview?: ContentOverview
  contentType?: ContentType
}

export interface GenerateTopicsResponse {
  title: string
  topics: TopicInput[]
}

export interface ChatMessageRequest {
  message: string
  phase: string
  topic?: string
  goal?: string
  topics?: TopicInput[]
  messages?: Message[]
}

export interface ChatMessageResponse {
  intent: string
  response: string
  data?: {
    title?: string
    topics?: TopicInput[]
  }
}

const API_BASE = 'http://localhost:8000'

async function fetchApi<T>(url: string, options?: RequestInit): Promise<T> {
  const response = await fetch(`${API_BASE}${url}`, {
    ...options,
    headers: {
      'Content-Type': 'application/json',
      ...options?.headers,
    },
  })
  
  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: 'Unknown error' }))
    throw new Error(error.detail || 'API error')
  }
  
  return response.json()
}

export const generateApi = {
  generateTopics: (topic: string, goal: string, messages?: Message[]) =>
    fetchApi<GenerateTopicsResponse>('/api/generate-topics', {
      method: 'POST',
      body: JSON.stringify({ topic, goal, messages }),
    }),

  chatMessage: (request: ChatMessageRequest) =>
    fetchApi<ChatMessageResponse>('/api/chat-message', {
      method: 'POST',
      body: JSON.stringify(request),
    }),

  generateTopicsFromChat: (messages: Message[], contentType: ContentType) =>
    fetchApi<GenerateTopicsResponse>('/api/generate-topics-from-chat', {
      method: 'POST',
      body: JSON.stringify({ messages, content_type: contentType }),
    }),
}
